File: Stack.c
/* Stack implementation using Array*/
#include<stdio.h>
#include<conio.h>
#include<stdlib.h>
#include<windows.h>
#define MAX 2 //Global variable
int stack[MAX]; //To create a stack using array of the MAX size
int top=-1; //It is declared to point the top of the stack and is set to -1 initially to denote that the stack is empty at starting


int isFull() //This function will check whether the stack index is not exceeding its max size
{
	if(top==MAX-1)
		return 1; //Can be interpreted as true and means that stack is full
	else
		return 0; //Can be interpreted as false and means that stack is not full
}

int isEmpty() //This function is used to check whether the stack is empty or not.
{
	if(top==-1)
		return 1; //Can be interpreted as true and means that stack is empty.
	else
		return 0; //Can be interpreted as false and means that stack is not empty.
}
void push() //Function to insert/push data into the stack
{
	int n;
	if(isFull())
	{
		printf("\n\n\aStack Overflow, can't add more data\n\n");
		printf("\n\nYou can try again following\n\n");
		return; //This will return to main or we can say that further code of this function will not get executed
	}
	printf("\nEnter the value to be pushed: ");
	scanf("%d",&n);
	top=top+1;  //First we will increase the top value
	stack[top]=n; //This will insert the data in the top of the stack
}

void pop() //Function to remove/pop the data from the stack
{
	int n; //To store the deleted item or the item which is in top
	if(isEmpty()){
		printf("\n\n\aStack Underflow, no value to pop..\n\n");
	}
	else{
		n=stack[top]; //To store the top item of the stack
		top=top-1; 
		printf("\n\nPop is successfull !, %d is deleted from stack\n\n",n);
	}
	
}
int peek() //To see the top element on the stack
{
	if(isEmpty())
	{
		printf("\n\n\aStack is Empty, No data to display\n\n");
	}
	else{
		printf("\n\nThe top element of the stack is %d\n\n",stack[top]); //shows the top element
	}
}
void print() //To see the list of all the elements present in the stack
{
	int i;
	if(isEmpty())
	{
		printf("\n\n\aStack Underflow, no data to show\n\n");
	}
	else
	{
		printf("\n\nThe elements in stack are:\n");
		for(i=top;i>=0;i--) //Loop is used to traverse all the elements of the stack
		{
			printf("%d\n",stack[i]);
		}
	}	
}

//Main function is a menu driven which provide the interface to user to perform all kinds of possible operations for stack
int main()
{
	int ch,value;
	while(1)
	{
		printf("\n");
		printf("_____________________________________ Stack Implementation Using Array __________________________________________\n\n");
		printf("1. Push");
		printf("\n2. Pop");
		printf("\n3. Peek");
		printf("\n4. Print all the elements of the stack");
		printf("\n5. Exit\n");
		printf("\nEnter your choice: ");
		scanf("%d",&ch);
		switch(ch)
		{
			case 1:
				push();
				break;
			case 2:
				pop();
				break;
			case 3:
				peek();
				break;
			case 4:
				print();
				break;
			case 5:
				exit(1);
			default:
				printf("Invalid choice\n");
		}		
	}
	return 0;
	getch();
}

File: Stack.py
# Menu driven program for stack implementation using list(array):-
class Stack:
    def __init__(self, max_size=100):
        self.stack = []  # Initialize an empty stack
        self.max_size = max_size  # Define maximum size of the stack

    def push(self, value):
        if len(self.stack) >= self.max_size:
            print("Stack Overflow! Cannot push", value)
        else:
            self.stack.append(value)
            print(f"{value} pushed onto the stack.")

    def pop(self):
        if not self.stack:
            print("Stack Underflow! No elements to pop.")
        else:
            value = self.stack.pop()
            print(f"{value} popped from the stack.")

    def peek(self):
        if not self.stack:
            print("Stack is empty. No top element.")
        else:
            print(f"Top element: {self.stack[-1]}")

    def display(self):
        if not self.stack:
            print("Stack is empty. No elements to display.")
        else:
            print("Stack elements:", self.stack[::-1])

    def is_empty(self):
        return len(self.stack) == 0

    def is_full(self):
        return len(self.stack) == self.max_size


def main():
    stack = Stack(max_size=5)  # Create a stack with a max size of 5
    while True:
        print("\n--- Stack Menu ---")
        print("1. Push")
        print("2. Pop")
        print("3. Peek")
        print("4. Display")
        print("5. Check if Empty")
        print("6. Check if Full")
        print("7. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            value = int(input("Enter value to push: "))
            stack.push(value)
        elif choice == 2:
            stack.pop()
        elif choice == 3:
            stack.peek()
        elif choice == 4:
            stack.display()
        elif choice == 5:
            print("Stack is empty." if stack.is_empty() else "Stack is not empty.")
        elif choice == 6:
            print("Stack is full." if stack.is_full() else "Stack is not full.")
        elif choice == 7:
            print("Exiting program.")
            break
        else:
            print("Invalid choice. Please try again.")


if __name__ == "__main__":
    main()

# # Explanation:-
# 1. In the main function, we create a stack with a max size of 5.
# 2. We use a while loop to provide a menu-driven interface to the user.
# 3. The user can choose from the options to push, pop, peek, display, check if empty, check if full, or exit the program.
# 4. The push option allows the user to enter a value and push it onto the stack.
# 5. The pop option removes the top element from the stack.
# 6. The peek option displays the top element of the stack.
# 7. The display option displays all the elements of the stack.
# 8. The is_empty and is_full methods check if the stack is empty or full, respectively.
# 9. The main function calls the menu-driven interface until the user chooses to exit.
# 10. The program ends when the user chooses to exit.


# Initialization:

# A stack is represented using a Python list.
# max_size is used to prevent stack overflow.

# Menu Options:

# Each operation (push, pop, peek, etc.) is handled based on user input.

# Edge Cases:

# Prevent pushing into a full stack (stack overflow).
# Prevent popping from an empty stack (stack underflow).